app/com/linkedin/drelephant/spark/legacydata/SparkJobProgressData.java

```diff
@@ -0,0 +1,273 @@
+  }
+
+  public void addFailedStages(int stageId, int attemptId) {
+    _failedStages.add(new StageAttemptId(stageId, attemptId));
+  }
+
+  public Set<Integer> getJobIds() {
+    return _jobIdToInfo.keySet();
+  }
+
+  public Set<StageAttemptId> getStageAttemptIds() {
+    return _stageIdToInfo.keySet();
+  }
+
+  public Set<Integer> getCompletedJobs() {
+    return _completedJobs;
+  }
+
+  public Set<Integer> getFailedJobs() {
+    return _failedJobs;
+  }
+
+  private static double getFailureRate(int numCompleted, int numFailed) {
+    int num = numCompleted + numFailed;
+
+    if (num == 0) {
+      return 0d;
+    }
+
+    return numFailed * 1.0d / num;
+  }
+
+  public double getJobFailureRate() {
+    return getFailureRate(_completedJobs.size(), _failedJobs.size());
+  }
+
+  public double getStageFailureRate() {
+    return getFailureRate(_completedStages.size(), _failedStages.size());
+  }
+
+  public JobInfo getJobInfo(int jobId) {
+    return _jobIdToInfo.get(jobId);
+  }
+
+  public StageInfo getStageInfo(int stageId, int attemptId) {
+    return _stageIdToInfo.get(new StageAttemptId(stageId, attemptId));
+  }
+
+  public Set<StageAttemptId> getCompletedStages() {
+    return _completedStages;
+  }
+
+  public Set<StageAttemptId> getFailedStages() {
+    return _failedStages;
+  }
+
+  /**
+   * Job itself does not have a name, it will use its latest stage as the name.
+   *
+   * @param jobId
+   * @return
+   */
+  public String getJobDescription(int jobId) {
+    List<Integer> stageIds = _jobIdToInfo.get(jobId).stageIds;
+    int id = -1;
+    for (int stageId : stageIds) {
+      id = Math.max(id, stageId);
+    }
+    if (id == -1) {
+      logger.error("Spark Job id [" + jobId + "] does not contain any stage.");
+      return null;
+    }
+    return _stageIdToInfo.get(new StageAttemptId(id, 0)).name;
+  }
+
+  public List<String> getFailedJobDescriptions() {
+    List<String> result = new ArrayList<String>();
+    for (int id : _failedJobs) {
+      result.add(getJobDescription(id));
+    }
+    return result;
+  }
+
+  // For debug purpose
+  public String toString() {
+    StringBuilder s = new StringBuilder();
+    s.append("JobInfo: [");
+
+    for (Map.Entry<Integer, JobInfo> entry : _jobIdToInfo.entrySet()) {
+      s.append("{id:" + entry.getKey() + ", value: " + entry.getValue() + "}");
+    }
+
+    s.append("]\nStageInfo: [");
+    for (Map.Entry<StageAttemptId, StageInfo> entry : _stageIdToInfo.entrySet()) {
+      s.append("{id:" + entry.getKey() + ", value: " + entry.getValue() + "}");
+    }
+    s.append("]");
+
+    return s.toString();
+  }
+
+  public static class StageAttemptId {
+    public int stageId;
+    public int attemptId;
+
+    public StageAttemptId(int stageId, int attemptId) {
+      this.stageId = stageId;
+      this.attemptId = attemptId;
+    }
+
+    @Override
+    public int hashCode() {
+      return new Integer(stageId).hashCode() * 31 + new Integer(attemptId).hashCode();
```

Comment:
Use `Ints.hashCode` from Guava instead of allocating a new `Integer` on each call, or just the `int` value as is.